Implement derivatives for nextafter operation

baskargopinath opened the pull request:
# Implement derivatives for nextafter operation

## Description
This PR implements the derivative for the `nextafter` operation, fixing issue #148814.

The derivative of `nextafter(x, y)` with respect to `x` is:
- 1.0 where x != y (since small changes in x result in proportional changes in output)
- 0.0 where x == y (since the function is constant at that point)

The derivative with respect to `y` is always zero.

## How Has This Been Tested?
I've tested this implementation with:
- Eager mode gradient calculation
- torch.compile (Inductor backend)
- Added test functionality to verify correct gradients are computed

Tests verify that gradients are correctly computed, with x getting gradients only where x != y.

## Fixes
Fixes #148814

baskargopinath committed:
Implement derivatives for nextafter operation

baskargopinath commented:
@pytorchbot label "topic: not user facing"

soulitzer commented on test/test_autograd.py:
```diff
@@ -990,6 +990,27 @@ def check_matches(out, inp):
+        expected_y_grad = torch.zeros(3)
+        
+        assert torch.allclose(x.grad, expected_x_grad)
+        assert torch.allclose(y.grad, expected_y_grad)
```
We have a OpInfo entry for `nextafter` in `torch/testing/_internal/common_methods_invocations.py` that we can update:

```
    BinaryUfuncInfo('nextafter',
                    dtypes=floating_types_and(torch.bfloat16, torch.half),
                    supports_autograd=False,
                    supports_rhs_python_scalar=False),
```

This should automatically add testing for gradients

baskargopinath replied:
resolved in https://github.com/pytorch/pytorch/pull/148820/commits/d9160045f2b65160441d23c59c69345e2e5da850

baskargopinath replied:
do u want me to remove this from test_autograd?

baskargopinath replied:
okay

soulitzer commented on torch/csrc/autograd/FunctionsManual.cpp:
```diff
@@ -177,6 +177,24 @@ static Tensor handle_r_to_c(const Tensor& self, Tensor gradient_result) {
+  // - 0.0 where x == y (no change in output when x equals y)
+  
+  auto mask = self.ne(other);
+  return grad * mask.to(grad.dtype());
```
You can avoid a .to if you just do torch.where

soulitzer commented on torch/csrc/autograd/FunctionsManual.cpp:
```diff
@@ -177,6 +177,24 @@ static Tensor handle_r_to_c(const Tensor& self, Tensor gradient_result) {
   return gradient_result;
 }
 
+Tensor nextafter_backward(const Tensor& grad, const Tensor& self, const Tensor& other) {
```
I would prefer this to be just inline in derivatives.yaml

soulitzer reviewed (commented):
Thanks! Added some small comments

baskargopinath committed:
Implement nextafter derivatives inline in YAML and update OpInfo

soulitzer commented on tools/autograd/derivatives.yaml:
```diff
@@ -1297,8 +1297,9 @@
 - name: nextafter(Tensor self, Tensor other) -> Tensor
-  self: not_implemented("nextafter")
-  other: not_implemented("nextafter")
+  self: torch.where(self != other, grad, torch.zeros_like(grad))
```
```suggestion
  self: at::where(self != other, grad, 0)
```
Sorry for the confusion here from my other comment. You can't use the python API (torch.*) because this file will eventually be used by cpp codegen. Also, we can avoid materializing the zeros tensor.

soulitzer commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -16576,7 +16576,7 @@ def sample_inputs_alias_copy(op_info, device, dtype, requires_grad, **kwargs):
     BinaryUfuncInfo('nextafter',
                     dtypes=floating_types_and(torch.bfloat16, torch.half),
-                    supports_autograd=False,
+                    supports_autograd=True,
```
```suggestion
                    supports_forward_ad=True,
                    supports_fwgrad_bwgrad=True,
```
supports_autograd is already by default True, so we don't need it anymore
we should also enable testing for forward AD though since we've added the jvp
I also expect  forward over backward to be supported automatically

baskargopinath committed:
Update tools/autograd/derivatives.yaml

Co-authored-by: Jeffrey Wan <soulitzer@gmail.com>

baskargopinath committed:
Update test_autograd and common_methods

soulitzer commented on tools/autograd/derivatives.yaml:
```diff
@@ -1297,8 +1297,9 @@
-  other: not_implemented("nextafter")
+  self: at::where(self != other, grad, 0)
+  other: zeros_like(other)
+  result: torch.where(self_p != other_p, self_t, torch.zeros_like(self_t))
```
```suggestion
  result: at::where(self_p != other_p, self_t, 0)
```

baskargopinath committed:
Update derivatives.yaml

Co-authored-by: Jeffrey Wan <soulitzer@gmail.com>

baskargopinath committed:
Update test_autograd.py

baskargopinath committed:
Update test_nestedtensor

baskargopinath commented:
@pytorchbot retest this please

baskargopinath commented:
I've fixed some failing tests for the nextafter derivatives PR, and here's what's changed:

1. **Updated `test_autograd.py`:**
   - Switched from `torch.nextafter` to `torch.pdist` in the `test_not_implemented_grad` test.
   - Updated the input tensor to 2D (`torch.rand(5, 2, ...)`) since that's what `pdist` requires.
   - Added a comment noting we chose `p=0.5` specifically to hit the `_pdist_backward` path, which intentionally lacks gradient implementation. This way, the test still effectively checks PyTorch's handling of missing gradient implementations.

2. **Updated `test_nestedtensor.py`:**
   - Added a SkipRule for `nextafter` to the `BACKWARD_SKIPS_AND_XFAILS` list:
   ```python
   SkipRule(
       op_match_fn=lambda device, op: op.full_name == "nextafter",
       name="nextafter_backward_not_implemented",
   ),
   ```
   - This temporarily skips backward tests for nested tensors with `nextafter` until proper support can be implemented.

baskargopinath committed:
Update test_autograd.py

baskargopinath commented:
I'd like to provide an update on the test_not_implemented_grad test issue:

## Issue
While implementing derivatives for `nextafter`, I discovered that `test_autograd.py::TestAutograd::test_not_implemented_grad` was failing with the error:
`AssertionError: NotImplementedError not raised`

## Root Cause Analysis
I found a significant discrepancy between derivatives.yaml and the actual implementations:

1. The test was expecting `torch.pdist(a, p=0.5)` to raise NotImplementedError during backward()
2. In derivatives.yaml, _pdist_backward is explicitly marked as `not_implemented("_pdist_backward")` for all inputs
3. However, when running the test, no error was raised and a functioning `SumBackward0` object was created

**This indicates a systemic issue**: Some functions marked as not_implemented in derivatives.yaml actually have working implementations in the codebase.

I encountered similar behavior with other functions like embedding_bag and histc - all marked as not_implemented in derivatives.yaml but functioning when tested.

## Fix
I adjusted the test to maintain its original purpose by:

1. Keeping the use of `torch.pdist` with `p=0.5`
2. Updating the input tensor shape to 2D (necessary for pdist)
3. Adding a comment explaining we're deliberately testing this specific path

## Broader Implications
The derivatives.yaml file needs to be audited and updated to reflect the actual state of gradient implementations. This will prevent future test confusion and ensure our documentation accurately reflects the codebase.

## Nextafter Derivatives Implementation
For the original PR, nextafter operation derivatives are implemented as follows:
- Derivative w.r.t. x is 1.0 where x != y (since changes in x yield proportional output changes)
- Derivative is 0.0 where x == y (since the function is constant there)
- Derivative w.r.t. y is always zero

baskargopinath committed:
Fix whitespace and formatting issues in test_autograd.py

baskargopinath commented:
@soulitzer can u try running the ci again

soulitzer commented on test/test_autograd.py:
```diff
@@ -547,9 +547,23 @@ def my_function(x, y):
+        y = CustomFunctionWithoutGrad.apply(x).sum()
+
+        print(f"Backward function: {y.grad_fn}")
+
```
Let's do
```
y = torch.acosh_(a.clone()).sum()
```
instead

What this test is testing is that the codegen'd autograd for built-in functions raises this error. (Maybe its worth improving the name of the test or adding a comment.)

baskargopinath replied:
done

baskargopinath committed:
Update test_autograd

baskargopinath committed:
Remove blankspace

Lucaskabela reviewed (dismissed):
Please update the SkipRule to XFailRule :) Other than that this LGTM

Skylion007 commented on test/test_nestedtensor.py:
```diff
@@ -8769,6 +8769,10 @@ def f(values, offsets):
+    SkipRule(
+        op_match_fn=lambda device, op: op.full_name == "nextafter",
+        name="nextafter_backward_not_implemented",
+    ),
```
```suggestion
    XFailRule(
        error_type=NotImplementedError,
        op_match_fn=lambda device, op: op.full_name == "nextafter",
        name="nextafter_backward_not_implemented",
    ),

Skylion007 commented:
@pytorcbot rebase

Skylion007 commented:
@pytorchbot rebase

Skylion007 committed:
Change SkipRule to XFail

Skylion007 commented on test/test_nestedtensor.py:
```diff
@@ -8769,6 +8769,11 @@ def f(values, offsets):
         name="unimplemented_masked_fill",
     ),
+    XFailRule(
+        error_type=NotImplementedError,
```
```suggestion
```

Skylion007 committed:
Broad exception xfail test/test_nestedtensor.py

Skylion007 commented:
@pytorchbot rebase

Skylion007 committed:
See what fails

Skylion007 committed:
Try to fix XFailRule

Skylion007 committed:
Fix sample_match_fn lambda for XFailRule

Skylion007 commented:
@pytorchbot merge

Skylion007 commented:
==================================== RERUNS ====================================
_______ TestConsistencyMPS.test_output_grad_match_nextafter_mps_float32 ________
Unexpected success
_______ TestConsistencyMPS.test_output_grad_match_nextafter_mps_float32 ________
Unexpected success
=================================== FAILURES ===================================
_______ TestConsistencyMPS.test_output_grad_match_nextafter_mps_float32 ________
Unexpected success
- generated xml file: /Users/ec2-user/runner/_work/pytorch/pytorch/test/test-reports/python-pytest/test_mps/test_mps-fd1ae416e96c29ee.xml -

Skylion007 committed:
Remove xfail nextafter MPS

Skylion007 commented:
@pytorchbot merge

Skylion007 commented:
@pytorchbot merge
